sdk/nodejs/policy/policy.ts

```diff
@@ -121,42 +121,29 @@ export interface ResourceValidationArgs {
 
 /**
  * A helper function that returns a strongly-typed resource validation function.
- * @param typeFilter A type guard used to determine if the args are an instance of the resource.
+ * @param resourceClass A resource class used to filter this check to only resources of the specified class and
+ * determine the appropriate strongly-typed `TArg` type to use for the resource.
  * @param validate A callback function that validates if the resource definition violates a policy.
  */
-export function validateTypedResource<TResource extends Resource>(
-    typeFilter: (o: any) => o is TResource,
+export function validateTypedResource<TResource extends Resource, TArgs>(
+    resourceClass: { new(name: string, args: TArgs, ...rest: any[]): TResource },
     validate: (
-        resource: q.ResolvedResource<TResource>,
+        props: NonNullable<Unwrap<TArgs>>,
```

Comment:
The resource's `TArgs` constructor parameter is typically optional, so use TypeScript's `NonNullable<T>` it isn't `TArgs | undefined` here. We also use [`Unwrap<T>`](https://github.com/pulumi/pulumi/blob/1a6897ba262c0668340fa5e62a26f89cc538ec41/sdk/nodejs/output.ts#L460-L490) so that the properties are typed as the underlying type instead of `Input<T>` (e.g. `string` instead of `Input<string>`.